debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/StringToJSON.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb.transforms;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.transforms.ExtractField;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * Debezium Mongo Connector generates the CDC records in String format. Sink connectors usually are not able to parse
+ * the string and insert the document as it is represented in the Source. so a user use this SMT to parse the String
+ * and insert the MongoDB document in the JSON format.
+ *
+ * @param <R> the subtype of {@link ConnectRecord} on which this transformation will operate
+ * @author Sairam Polavarapu
+ */
+public class StringToJSON<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    private final ExtractField<R> afterExtractor = new ExtractField.Value<R>();
+    private final ExtractField<R> patchExtractor = new ExtractField.Value<R>();
+    private final ExtractField<R> keyExtractor = new ExtractField.Key<R>();
+
+    @Override
+    public R apply(R r) {
+        SchemaBuilder schemabuilder = SchemaBuilder.struct();
+        SchemaBuilder schemabuilder1 = SchemaBuilder.struct();
+        BsonDocument value = null;
+        BsonDocument Key = null;
+        ObjectMapper mapper = new ObjectMapper();
+
+        final R afterRecord = afterExtractor.apply(r);
+
+        if (afterRecord.value() == null) {
+            final R patchRecord = patchExtractor.apply(r);
+            final R key = keyExtractor.apply(r);
+            ObjectNode patchEventWKey = mapper.createObjectNode();
+            JsonNode patchEvent = null;
+
+            try {
```

Comment:
Why is it that you first parse using Jackson here and then further below serialize the object and parse it again using `BsonDocument`?

Maybe you can describe what you actually would like to achieve here? I think this code can be much simpler. Also we should avoid `printStackTrace()`, rather wrap and rethrow e.g. in a `ConnectException`. But ideally we'd not need the parsing step using Jackson at all.

Reply:
So I think this is what should be happening in the patch case essentially (plus probably some handling in case $set isn't there):

```java
        if (afterRecord.value() == null) {
            final R patchRecord = patchExtractor.apply(r);
            final R key = keyExtractor.apply(r);

            value = BsonDocument.parse(patchRecord.value().toString());
            value = value.getDocument("$set");
            Key = BsonDocument.parse("{ \"id\" : " + key.key().toString() + "}");

            if (!value.containsKey("_id")) {
                value.append("_id", Key.get("id"));
            }
        }
```